BlueprintUI/Sources/Element/MeasurementCachingKey.swift

```diff
@@ -0,0 +1,114 @@
+/// of your element **must** be included in the `input` key.
+///
+/// Caching is only done on a per-layout pass. Caches are not reused between layout passes.
+///
+/// Example
+/// -------
+/// Below is a simplified version of a `MeasurementCachingKey` in action, in this case for Blueprint's `Label` type.
+/// ```
+/// struct Label : UIViewElement {
+///
+///     var text : String
+///     var font : UIFont
+///     var numberOfLines : Int
+///
+///     typealias UIViewType = UILabel
+///
+///     static func makeUIView() -> UILabel {
+///         UILabel()
+///     }
+///
+///     func updateUIView(_ view : UILabel) {
+///         view.text = self.text
+///         view.font = self.font
+///         view.numberOfLines = self.numberOfLines
+///     }
+///
+///     var measurementCacheKey : AnyHashable? {
+///         MeasurementCachingKey(
+///             type: Self.self,
+///             input: Key(
+///               text: self.text,
+///               font: self.font,
+///               numberOfLines: self.numberOfLines
+///             )
+///         )
+///     }
+///
+///     struct Key : Hashable {
+///         var text : String
+///         var font : UIFont
+///         var numberOfLines : Int
+///     }
+/// }
+/// ```
+/// You may notice that the `Key`'s stored properties are identical to the stored properties of the Label type itself.
+/// When this occurs, you can simplify your implementation like this, by making the `Element` conform to `Hashable`.
+///
+/// ```
+/// struct Label : UIViewElement, Hashable {
+///
+///     var text : String
+///     var font : UIFont
+///     var numberOfLines : Int
+///
+///     typealias UIViewType = UILabel
+///
+///     static func makeUIView() -> UILabel {
+///         UILabel()
+///     }
+///
+///     func updateUIView(_ view : UILabel) {
+///         view.text = self.text
+///         view.font = self.font
+///         view.numberOfLines = self.numberOfLines
+///     }
+///
+///     var measurementCacheKey : AnyHashable? {
+///         MeasurementCachingKey(
+///         type: Self.self,
+///         input: self
+///     }
+/// }
+/// ```
+public struct MeasurementCachingKey : Hashable
+{
+    private let elementType : ObjectIdentifier
+    private let input : AnyHashable
+    
+    private let hash : Int
+    
+    public init<ElementType:Element, Input:Hashable>(type : ElementType.Type, input : Input)
+    {
+        self.elementType = ObjectIdentifier(type)
+        self.input = input
+        
+        var hasher = Hasher()
+        hasher.combine(self.elementType)
+        hasher.combine(self.input)
+        self.hash = hasher.finalize()
+    }
+    
+    public func hash(into hasher: inout Hasher) {
+        // We pre-compute the hash above to avoid potentially expensive work here.
+        // Computing the hash repeatedly for `input` may be slow (especially for strings).
+        hasher.combine(self.hash)
```

Comment:
I see MeasurementCache.Key also does this — do we need both?

Reply:
Ooh good call; yeah we only need this one, I think...